onnxruntime/contrib_ops/rocm/math/gemm_float8.cu

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/common/common.h"
+#include "core/framework/float16.h"
+#include "core/providers/rocm/rocm_kernel.h"
+#include "contrib_ops/rocm/math/gemm_float8_ck.cuh"
+
+namespace onnxruntime {
+namespace contrib {
+namespace rocm {
+
+using namespace onnxruntime::rocm;
+using namespace onnxruntime::rocm::tunable::blas;
+
+class GemmFloat8 final : public RocmKernel {
+ public:
+  GemmFloat8(const OpKernelInfo& info) : RocmKernel(info) {
+    transA_ = info.GetAttrOrDefault<int64_t>("transA", 0);
+    transB_ = info.GetAttrOrDefault<int64_t>("transB", 0);
+    dtype_ = info.GetAttrOrDefault<int64_t>("dtype", onnx::TensorProto_DataType_FLOAT16);
+    alpha_ = info.GetAttrOrDefault<float>("alpha", 1);
+    beta_ = info.GetAttrOrDefault<float>("beta", 0);
+  }
+  Status ComputeInternal(OpKernelContext* ctx) const override;
+
+ private:
+#if !defined(DISABLE_FLOAT8_TYPES)
+  template <typename Fp8T>
+  Status ComputeFp8Fp16Fp16(OpKernelContext* ctx, const Tensor* A, const Tensor* scaleA, const Tensor* B, Tensor* C) const;
+  template <typename Fp8T>
+  Status ComputeFp16Fp8Fp16(OpKernelContext* ctx, const Tensor* A, const Tensor* B, const Tensor* scaleB, Tensor* C) const;
+
+  template <typename TA, typename TB, typename TC, BlasOp OpA, BlasOp OpB>
+  [[nodiscard]] inline auto* GetOp() const {
+    using OpT = GemmFloat8TunableOp<TA, TB, TC, OpA, OpB>;
+    if (tunable_op_) {
+      return static_cast<OpT*>(tunable_op_.get());
+    }
+
+    auto create = std::make_unique<OpT>();   // avoid new
+    tunable_op_ = std::shared_ptr<void>(create.release(), [](void* ptr) {
+      auto release = std::unique_ptr<OpT>();  // avoid delete
+      release.reset(static_cast<OpT*>(ptr));
+    });
+
+    return static_cast<OpT*>(tunable_op_.get());
+  }
+#endif
+
+  float alpha_;
+  float beta_;
+  bool transA_;
+  bool transB_;
+  int64_t dtype_;
+
+  // fully type erased
+  mutable std::shared_ptr<void> tunable_op_;
+};
+
+Status GemmFloat8::ComputeInternal(OpKernelContext* ctx) const {
+#if defined(DISABLE_FLOAT8_TYPES)
+  return ORT_MAKE_STATUS(ONNXRUNTIME, FAIL, "DISABLE_FLOAT8_TYPES");
+#else
+  const Tensor* A = ctx->Input<Tensor>(0);
+  const Tensor* B = ctx->Input<Tensor>(1);
+  const Tensor* C = ctx->Input<Tensor>(2);  // bias
+  const Tensor* scale_a = ctx->Input<Tensor>(3);
+  const Tensor* scale_b = ctx->Input<Tensor>(4);
+  const Tensor* scale_y = ctx->Input<Tensor>(5);
+
+  auto a_shape = A->Shape();
+  auto b_shape = B->Shape();
+  ORT_ENFORCE(a_shape.NumDimensions() >= 2 && b_shape.NumDimensions() == 2);  // is in form of input @ weight
+  ORT_ENFORCE(a_shape[a_shape.NumDimensions() - 1] == b_shape[0]);            // k is compatiable
```

Comment:
Is b_shape fixed at kxn? I guess some code below about b_shape assumes that b_shape is kxn, but B transpose is supported.